File: cses/dynamic-programming/cc/book_shop.cc
#include <iostream>
#include <vector>

int main() {
    int n, x;
    std::cin >> n >> x;
    std::vector<int> prices(n), pages(n);

    for (int i = 0; i < n; i++)
        std::cin >> prices[i];

    for (int i = 0; i < n; i++)
        std::cin >> pages[i];

    std::vector<std::vector<int>> mem(n+1, std::vector<int>(x+1, 0));

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= x; j++) {
            mem[i][j] = mem[i-1][j];

            if (j >= prices[i-1])
                mem[i][j] = std::max(mem[i][j], mem[i-1][j-prices[i-1]] + pages[i-1]);
        }
    }

    std::cout << mem[n][x] << '\n';
    return 0;
}

File: cses/dynamic-programming/java/bookShop.java
import java.util.Scanner;

public class bookShop {
    public static void main(String[] a) {
        solve();
    }

    static void solve() {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int x = sc.nextInt();
        int[] prices = new int[n];
        int[] pages = new int[n];

        for (int i = 0; i < n; i++) {
            prices[i] = sc.nextInt();
        }

        for (int i = 0; i < n; i++) {
            pages[i] = sc.nextInt();
        }
        sc.close();

        int[][] mem = new int[n+1][x+1];

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= x; j++) {
                mem[i][j] = mem[i-1][j];

                if (j >= prices[i-1])
                    mem[i][j] = Math.max(mem[i][j], mem[i-1][j-prices[i-1]] + pages[i-1]);
            }
        }

        System.out.println(mem[n][x]);
    }
}
